compiler/rustc_mir_build/src/check_unsafety.rs

```diff
@@ -246,11 +292,86 @@ impl<'a, 'tcx> Visitor<'a, 'tcx> for UnsafetyVisitor<'a, 'tcx> {
                 // Unsafe blocks can be used in closures, make sure to take it into account
                 self.safety_context = closure_visitor.safety_context;
             }
+            ExprKind::Borrow { borrow_kind, arg } => {
+                let mut visitor = FieldVisitor::new(self.thir, self.tcx);
+                visit::walk_expr(&mut visitor, expr);
+
+                if visitor.found {
+                    match borrow_kind {
+                        BorrowKind::Shallow | BorrowKind::Shared | BorrowKind::Unique
+                            if !self.thir[arg]
+                                .ty
+                                .is_freeze(self.tcx.at(self.thir[arg].span), self.param_env) =>
+                        {
+                            self.requires_unsafe(expr.span, BorrowOfLayoutConstrainedField);
+                        }
+                        BorrowKind::Mut { .. } => {
+                            self.requires_unsafe(expr.span, MutationOfLayoutConstrainedField);
+                        }
+                        _ => {}
+                    }
+                }
+            }
+            ExprKind::Assign { lhs, .. } | ExprKind::AssignOp { lhs, .. } => {
+                let mut visitor = FieldVisitor::new(self.thir, self.tcx);
+                visit::walk_expr(&mut visitor, &self.thir[lhs]);
+                if visitor.found {
+                    self.requires_unsafe(expr.span, MutationOfLayoutConstrainedField);
+                }
+            }
             _ => {}
         }
 
         visit::walk_expr(self, expr);
     }
+
+    fn visit_pat(&mut self, pat: &Pat<'tcx>) {
+        match &*pat.kind {
+            PatKind::Variant { .. } | PatKind::Leaf { .. } => {
+                if let ty::Adt(adt_def, ..) = pat.ty.kind() {
+                    if (Bound::Unbounded, Bound::Unbounded)
+                        != self.tcx.layout_scalar_valid_range(adt_def.did)
+                    {
+                        let old_inside_adt = std::mem::replace(&mut self.inside_adt, true);
+                        visit::walk_pat(self, pat);
+                        self.inside_adt = old_inside_adt;
+                    }
```

Comment:
So I found this case which should error but didn't:
```rs
let mut foo = unsafe { NonZero((1,)) };
match &mut foo {
    NonZero((a,)) => *a = 0,
}
```
and I couldn't understand why it didn't catch that there was a `ref mut` in there. Turns out it's just missing a `walk_pat` here and I didn't see it :sweat_smile:
```suggestion
                    } else {
                        visit::walk_pat(self, pat);
                    }
```